zebra-consensus/src/verify.rs

```diff
@@ -1,60 +1,75 @@
-//! Context-dependent block and transaction verification for Zebra.
+//! Block verification and chain state updates for Zebra.
 //!
-//! Verification occurs in two stages:
-//!   - context-free verification of signatures, proofs, and scripts, handled by `zebra-chain` and
-//!     `zebra-script`
-//!   - context-dependent verification of the chain, handled by `zebra-consensus` (this module)
+//! Verification occurs in multiple stages:
+//!   - getting blocks (disk- or network-bound)
+//!   - context-free verification of signatures, proofs, and scripts (CPU-bound)
+//!   - context-dependent verification of the chain state (awaits a verified parent block)
 //!
-//! Verification is provided via a `tower::Service`, to support batching.
+//! Verification is provided via a `tower::Service`, to support backpressure and batch
+//! verification.
+
+use std::{
+    error::Error,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::{buffer::Buffer, Service};
 
-use std::sync::Arc;
 use zebra_chain::block::Block;
-use zebra_chain::transaction::Transaction;
+use zebra_state;
+
+mod transaction;
 
-/// A context-dependent block or transaction verification request.
+/// Block verification service.
 ///
-/// This has a `Flush` variant, to allow manually flushing queued verification
-/// requests, even when the batching service is wrapped in other `tower` layers.
-#[derive(Debug)]
-// The associated data doesn't need separate documentation.
-#[allow(missing_docs)]
-pub enum Request {
-    /// Verify the block and its transactions, and check them against the chain state
-    VerifyBlock { block: Arc<Block> },
-    /// Verify the transaction, and check it against the chain state
-    VerifyTransaction { transaction: Arc<Transaction> },
-    /// Flush incomplete verification batches
-    Flush,
+/// After verification, blocks and their associated transactions are added to
+/// `zebra-state::ZebraState`.
+#[derive(Default)]
+struct ZebraBlock {}
+
+/// The result type for the ZebraBlock Service.
+// TODO(teor): Response = BlockHeaderHash
+type ServiceResponse = ();
+
+/// The error type for the ZebraBlock Service.
+// TODO(teor): Error = Report ?
```

Comment:
```suggestion
// TODO(jane): Error = Report ?
```

heh, I'll handle this one, I am on the hook for the rest of the `Report` stuff for services too. The reason we're still using `Box<dyn Error...>` here and elsewhere is because `tower` hasn't implemented `map_err` for `Service` yet, so we cannot convert the return type of the Service for `Buffer` easily. We might just implement a `map_err`-like service specifically for mapping to a `Report` to do this short term but that's all stuff on my plate so yea.

Reply:
Switched the TODO to you in c620e4b42dc367ffd6fd8b20bfeade182eab7ff2.